labgrid/util/snmp.py

```diff
@@ -8,24 +10,39 @@ def __init__(self, host, community, port=161):
         if port is None:
             port = 161
 
+        try:
+            # if called from async code, try to get current's thread loop
+            self.loop = asyncio.get_running_loop()
+        except RuntimeError:
+            # no previous, external or running loop found, create a new one
+            self.loop = asyncio.new_event_loop()
+
         self.engine = hlapi.SnmpEngine()
-        self.transport = hlapi.UdpTransportTarget((host, port))
+        self.transport = self.loop.run_until_complete(hlapi.UdpTransportTarget.create((host, port)))
         self.community = hlapi.CommunityData(community, mpModel=0)
         self.context = hlapi.ContextData()
 
     def get(self, oid):
-        g = hlapi.getCmd(self.engine, self.community, self.transport,
+        g = self.loop.run_until_complete(hlapi.getCmd(self.engine, self.community, self.transport,
             self.context, hlapi.ObjectType(hlapi.ObjectIdentity(oid)),
-            lookupMib=False)
+            lookupMib=False))
 
-        error_indication, error_status, _, res = next(g)
+        error_indication, error_status, _, res = g
         if error_indication or error_status:
             raise ExecutionError("Failed to get SNMP value.")
         return res[0][1]
 
     def set(self, oid, value):
         identify = hlapi.ObjectType(hlapi.ObjectIdentity(oid),
                    hlapi.Integer(value))
-        g = hlapi.setCmd(self.engine, self.community, self.transport,
-            self.context, identify, lookupMib=False)
-        next(g)
+        g = self.loop.run_until_complete(hlapi.setCmd(self.engine, self.community, self.transport,
+            self.context, identify, lookupMib=False))
+
+        error_indication, error_status, _, _ = g
+        if error_indication or error_status:
+            raise ExecutionError("Failed to set SNMP value.")
+
+    def cleanup(self):
+        self.engine.closeDispatcher()
+        self.loop.run_until_complete(self.loop.shutdown_asyncgens())
+        self.loop.close()
```

Comment:
Doesn't this close a potential external loop?

Reply:
Ah, yes, we should probably stash the fact that we created a loop and only close this if we did.

Reply:
Implemented.